File: packages/twenty-apps/internal/twenty-partners/src/skills/twenty-partner-application-triage/rank.ts
// Deterministic ranker for the twenty-partner-application-triage skill.
//
// Ranks APPLICATION-stage partners by NET-NEW value vs the current VALIDATED set:
// geography and language we don't yet cover (weighted high), plus a "real Twenty
// work" proof signal read from the application notes. Skill volume is deliberately
// capped so generic dev shops that spray skill lists don't dominate.
//
// Reads creds from ~/.twenty/credentials.env. Emits ranked JSON to stdout.
// Run `tsx rank.ts --selftest` to verify scoring without hitting the API.
//
// ponytail: fixed weights + tier thresholds. Tune the WEIGHTS/THRESHOLDS dicts
// below if the ranking drifts; everything else is mechanical. Raw REST + own
// types on purpose: the codegen SDK schema only carries the Partner object after
// a local sync, so a typed client.query would not compile in a clean checkout.

import { readFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

const WEIGHTS = { geo: 3, lang: 3, scope: 1, skill: 1, skillCap: 3, proof: 6 };
const THRESHOLDS = { A: 12, B: 5 }; // >=A => A; >=B => B; else C. Any proof => at least A.

// "real Twenty work" signals in the free-text notes (the high-confidence axis).
const PROOF_WORKSPACE = /https?:\/\/[^\s]*(twenty|crm)[^\s]*/i;
const PROOF_CUSTOMERS = /(customers?\s+onboarded|real implementation|clients?\s+(moving|migrat)|delivered \d|named customer)/i;
const PROOF_MIGRATION = /(switch\w*|migrat\w*|moved|replac\w*|our crm|own crm|we use twenty|dogfood|managed service)/i;

const CRED_PATH = join(homedir(), '.twenty', 'credentials.env');
const HTTP_TIMEOUT_MS = 90_000;

type PersonRecord = {
  name?: { firstName?: string | null; lastName?: string | null } | null;
  emails?: { primaryEmail?: string | null } | null;
  linkedinLink?: unknown;
};

type PartnerRecord = {
  name?: string | null;
  validationStage?: string | null;
  region?: unknown;
  country?: unknown;
  languagesSpoken?: unknown;
  partnerScope?: unknown;
  skills?: unknown;
  typeOfTeam?: string | null;
  applicationNotes?: unknown;
  callBookedAt?: unknown;
  persons?: PersonRecord[] | null;
  linkedin?: unknown;
  website?: unknown;
};

type PartnersPage = {
  data?: { partners?: PartnerRecord[] };
  pageInfo?: { hasNextPage?: boolean; endCursor?: string };
};

const loadCreds = (): { url: string; key: string } => {
  let raw: string;
  try {
    raw = readFileSync(CRED_PATH, 'utf8');
  } catch {
    throw new Error(
      `Missing ${CRED_PATH}. Copy the partners key from the app's .env.prod into it ` +
        '(TWENTY_PARTNERS_API_URL / TWENTY_PARTNERS_API_KEY).',
    );
  }
  const env: Record<string, string> = {};
  for (const line of raw.split('\n')) {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#') && trimmed.includes('=')) {
      const idx = trimmed.indexOf('=');
      env[trimmed.slice(0, idx).trim()] = trimmed.slice(idx + 1).trim();
    }
  }
  const url = env.TWENTY_PARTNERS_API_URL;
  const key = env.TWENTY_PARTNERS_API_KEY;
  if (!url || !key) {
    throw new Error('credentials.env is missing TWENTY_PARTNERS_API_URL or TWENTY_PARTNERS_API_KEY.');
  }
  return { url: url.replace(/\/$/, ''), key };
};

const fetchAllPartners = async (url: string, key: string): Promise<PartnerRecord[]> => {
  const recs: PartnerRecord[] = [];
  let after: string | undefined;
  for (;;) {
    const path = `${url}/rest/partners?limit=60&depth=1${after ? `&starting_after=${after}` : ''}`;
    const res = await fetch(path, {
      headers: { Authorization: `Bearer ${key}`, 'User-Agent': 'Mozilla/5.0' },
      signal: AbortSignal.timeout(HTTP_TIMEOUT_MS),
    });
    if (!res.ok) {
      const hint =
        res.status === 401 || res.status === 403
          ? ' — auth failed, check TWENTY_PARTNERS_API_KEY in ~/.twenty/credentials.env'
          : '';
      throw new Error(`Partners API ${res.status} ${res.statusText}${hint}`);
    }
    const body = (await res.json()) as PartnersPage;
    const page = body.data?.partners ?? [];
    recs.push(...page);
    if (body.pageInfo?.hasNextPage && page.length > 0 && body.pageInfo.endCursor) {
      after = body.pageInfo.endCursor;
      continue;
    }
    return recs;
  }
};

// Flatten any nested value into a set of lowercased non-empty string tokens.
const tok = (value: unknown): Set<string> => {
  const out = new Set<string>();
  if (value === null || value === undefined) return out;
  if (typeof value === 'string') {
    const s = value.trim().toLowerCase();
    if (s) out.add(s);
  } else if (Array.isArray(value)) {
    for (const item of value) for (const t of tok(item)) out.add(t);
  } else if (typeof value === 'object') {
    for (const item of Object.values(value as Record<string, unknown>)) for (const t of tok(item)) out.add(t);
  } else {
    out.add(String(value).toLowerCase());
  }
  return out;
};

const diffSorted = (value: Set<string>, base: Set<string>): string[] =>
  [...value].filter((x) => !base.has(x)).sort();

const notesStr = (rec: PartnerRecord): string => {
  const v = rec.applicationNotes;
  if (typeof v === 'string') return v;
  return v === null || v === undefined ? '' : JSON.stringify(v);
};

const linkUrl = (value: unknown): string | null => {
  if (typeof value === 'string') return value;
  if (value && typeof value === 'object') {
    const u = (value as { primaryLinkUrl?: unknown }).primaryLinkUrl;
    return typeof u === 'string' ? u : null;
  }
  return null;
};

// Mirror Python truthiness so the linkedin `or` fallback ports exactly: "", 0,
// false, null/undefined, and EMPTY array/object are falsy (JS would treat {}/[] as
// truthy, which would diverge from rank.py's `p.linkedinLink or rec.linkedin`).
const pyTruthy = (v: unknown): boolean => {
  if (v === null || v === undefined) return false;
  if (typeof v === 'string') return v.length > 0;
  if (typeof v === 'number') return v !== 0;
  if (typeof v === 'boolean') return v;
  if (Array.isArray(v)) return v.length > 0;
  if (typeof v === 'object') return Object.keys(v).length > 0;
  return true;
};

const contact = (rec: PartnerRecord): { name: string | null; email: string | null; linkedin: string | null } => {
  const persons = rec.persons ?? [];
  if (persons.length === 0) return { name: null, email: null, linkedin: linkUrl(rec.linkedin) };
  const p = persons[0];
  const full = [p.name?.firstName, p.name?.lastName].filter((x): x is string => Boolean(x)).join(' ') || null;
  const email = p.emails?.primaryEmail ?? null;
  // Python: `p.get("linkedinLink") or rec.get("linkedin")`, then normalize. Choose
  // the raw value with Python falsiness FIRST, normalize once after.
  const rawLinkedin = pyTruthy(p.linkedinLink) ? p.linkedinLink : rec.linkedin;
  return { name: full, email, linkedin: linkUrl(rawLinkedin) };
};

type Baseline = { geo: Set<string>; lang: Set<string>; scope: Set<string>; skill: Set<string> };

const baseline = (validated: PartnerRecord[]): Baseline => {
  const base: Baseline = { geo: new Set(), lang: new Set(), scope: new Set(), skill: new Set() };
  for (const r of validated) {
    for (const t of tok(r.region)) base.geo.add(t);
    for (const t of tok(r.country)) base.geo.add(t);
    for (const t of tok(r.languagesSpoken)) base.lang.add(t);
    for (const t of tok(r.partnerScope)) base.scope.add(t);
    for (const t of tok(r.skills)) base.skill.add(t);
  }
  return base;
};

type RankedEntry = {
  name: string | null | undefined;
  score: number;
  tier: 'A' | 'B' | 'C';
  new_geo: string[];
  new_lang: string[];
  new_scope: string[];
  new_skills: string[];
  proof: Record<string, boolean>;
  team: string | null | undefined;
  contact_name: string | null;
  email: string | null;
  linkedin: string | null;
  website: string | null;
  notes: string;
};

const scoreOne = (rec: PartnerRecord, base: Baseline): RankedEntry => {
  const newGeo = diffSorted(new Set([...tok(rec.region), ...tok(rec.country)]), base.geo);
  const newLang = diffSorted(tok(rec.languagesSpoken), base.lang);
  const newScope = diffSorted(tok(rec.partnerScope), base.scope);
  const newSkill = diffSorted(tok(rec.skills), base.skill);
  const notes = notesStr(rec);
  const proof = {
    workspace_url: PROOF_WORKSPACE.test(notes),
    customers: PROOF_CUSTOMERS.test(notes),
    migration: PROOF_MIGRATION.test(notes),
  };
  const hasProof = Object.values(proof).some(Boolean);
  const score =
    WEIGHTS.geo * newGeo.length +
    WEIGHTS.lang * newLang.length +
    WEIGHTS.scope * newScope.length +
    WEIGHTS.skill * Math.min(newSkill.length, WEIGHTS.skillCap) +
    (hasProof ? WEIGHTS.proof : 0);
  const tier: 'A' | 'B' | 'C' = hasProof || score >= THRESHOLDS.A ? 'A' : score >= THRESHOLDS.B ? 'B' : 'C';
  const { name, email, linkedin } = contact(rec);
  return {
    name: rec.name,
    score,
    tier,
    new_geo: newGeo,
    new_lang: newLang,
    new_scope: newScope,
    new_skills: newSkill,
    proof: Object.fromEntries(Object.entries(proof).filter(([, v]) => v)),
    team: rec.typeOfTeam,
    contact_name: name,
    email,
    linkedin,
    website: linkUrl(rec.website),
    notes: notes.trim().slice(0, 400),
  };
};

type RankResult = {
  validated_count: number;
  application_count: number;
  booking_state_wired: boolean;
  coverage: Record<string, string[]>;
  ranked: RankedEntry[];
};

const rank = (recs: PartnerRecord[]): RankResult => {
  const validated = recs.filter((r) => r.validationStage === 'VALIDATED');
  let apps = recs.filter((r) => r.validationStage === 'APPLICATION');
  // Forward-compat: once callBookedAt exists, narrow to the un-booked (the chase set).
  const hasBookedField = recs.some((r) => 'callBookedAt' in r);
  if (hasBookedField) apps = apps.filter((r) => !r.callBookedAt);
  const base = baseline(validated);
  const ranked = apps.map((r) => scoreOne(r, base)).sort((a, b) => b.score - a.score);
  return {
    validated_count: validated.length,
    application_count: apps.length,
    booking_state_wired: hasBookedField,
    coverage: Object.fromEntries(Object.entries(base).map(([k, v]) => [k, [...v].sort()])),
    ranked,
  };
};

const assert = (cond: boolean, msg: string): void => {
  if (!cond) throw new Error(`selftest failed: ${msg}`);
};

const selftest = (): void => {
  const baseRecs: PartnerRecord[] = [
    {
      validationStage: 'VALIDATED',
      country: ['france'],
      languagesSpoken: ['french', 'english'],
      partnerScope: ['development'],
      skills: ['react', 'postgres'],
    },
  ];
  const apps: PartnerRecord[] = [
    {
      validationStage: 'APPLICATION',
      name: 'gap+proof',
      country: ['germany'],
      languagesSpoken: ['german'],
      applicationNotes: 'Live workspace https://crm.acme.de — customers onboarded: Foo GmbH',
    },
    {
      validationStage: 'APPLICATION',
      name: 'skill-sprayer',
      skills: ['php', 'vue', 'kotlin', 'swift', 'laravel', 'mongodb'],
    },
    { validationStage: 'APPLICATION', name: 'empty' },
  ];
  const out = rank([...baseRecs, ...apps]).ranked;
  const order = out.map((r) => r.name);
  assert(JSON.stringify(order) === JSON.stringify(['gap+proof', 'skill-sprayer', 'empty']), JSON.stringify(order));
  assert(out[0].tier === 'A', JSON.stringify(out[0]));
  assert(out[1].score === WEIGHTS.skillCap, JSON.stringify(out[1])); // 6 new skills capped at 3
  assert(out[2].tier === 'C', JSON.stringify(out[2]));

  // linkedin fallback parity with Python `or`: an empty person link falls back to
  // the record link; a person link dict with a null url does NOT (the dict is
  // truthy in Python, so it wins and normalizes to null).
  const recLink = { primaryLinkUrl: 'https://li/rec', secondaryLinks: [] };
  const emptyDict = rank([
    { validationStage: 'APPLICATION', name: 'li', persons: [{ linkedinLink: {} }], linkedin: recLink },
  ]).ranked[0];
  assert(emptyDict.linkedin === 'https://li/rec', `empty person link should fall back: ${emptyDict.linkedin}`);
  const nullUrlDict = rank([
    { validationStage: 'APPLICATION', name: 'li2', persons: [{ linkedinLink: { primaryLinkUrl: null } }], linkedin: recLink },
  ]).ranked[0];
  assert(nullUrlDict.linkedin === null, `null-url person link should not fall back: ${nullUrlDict.linkedin}`);

  console.log('selftest ok:', order);
};

const main = async (): Promise<void> => {
  if (process.argv.includes('--selftest')) {
    selftest();
    return;
  }
  const { url, key } = loadCreds();
  const result = rank(await fetchAllPartners(url, key));
  console.log(JSON.stringify(result, null, 2));
};

main().catch((err: unknown) => {
  console.error(err instanceof Error ? err.message : String(err));
  process.exit(1);
});

File: packages/twenty-apps/internal/twenty-partners/src/skills/twenty-partner-application-triage/rank.py
#!/usr/bin/env python3
"""Deterministic ranker for the twenty-partner-application-triage skill.

Ranks APPLICATION-stage partners by NET-NEW value vs the current VALIDATED set:
geography and language we don't yet cover (weighted high), plus a "real Twenty
work" proof signal read from the application notes. Skill volume is deliberately
capped so generic dev shops that spray skill lists don't dominate.

Reads creds from ~/.twenty/credentials.env. Emits ranked JSON to stdout.
Run `rank.py --selftest` to verify scoring without hitting the API.

# ponytail: fixed weights + tier thresholds. Tune the WEIGHTS/THRESHOLDS dicts
# below if the ranking drifts; everything else is mechanical.
"""
import json
import os
import re
import sys
import urllib.request

WEIGHTS = {"geo": 3, "lang": 3, "scope": 1, "skill": 1, "skill_cap": 3, "proof": 6}
THRESHOLDS = {"A": 12, "B": 5}  # >=A => A; >=B => B; else C. Any proof => at least A.

# "real Twenty work" signals in the free-text notes (the high-confidence axis).
PROOF_WORKSPACE = re.compile(r"https?://[^\s]*(twenty|crm)[^\s]*", re.I)
PROOF_CUSTOMERS = re.compile(r"(customers?\s+onboarded|real implementation|clients?\s+(moving|migrat)|delivered \d|named customer)", re.I)
PROOF_MIGRATION = re.compile(r"(switch\w*|migrat\w*|moved|replac\w*|our crm|own crm|we use twenty|dogfood|managed service)", re.I)

CRED_PATH = os.path.expanduser("~/.twenty/credentials.env")


def load_creds():
    if not os.path.exists(CRED_PATH):
        sys.exit(f"Missing {CRED_PATH}. Copy the partners key from the app's "
                 ".env.prod into it (TWENTY_PARTNERS_API_URL / _API_KEY).")
    env = {}
    with open(CRED_PATH) as fh:
        for line in fh:
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, v = line.split("=", 1)
                env[k.strip()] = v.strip()
    url = env.get("TWENTY_PARTNERS_API_URL")
    key = env.get("TWENTY_PARTNERS_API_KEY")
    if not url or not key:
        sys.exit("credentials.env is missing TWENTY_PARTNERS_API_URL or TWENTY_PARTNERS_API_KEY.")
    return url.rstrip("/"), key


def fetch_all_partners(url, key):
    recs, after = [], None
    while True:
        path = f"{url}/rest/partners?limit=60&depth=1" + (f"&starting_after={after}" if after else "")
        req = urllib.request.Request(path, headers={"Authorization": f"Bearer {key}", "User-Agent": "Mozilla/5.0"})
        data = json.load(urllib.request.urlopen(req))
        page = data["data"]["partners"]
        recs += page
        info = data.get("pageInfo", {})
        if info.get("hasNextPage") and page:
            after = info["endCursor"]
            continue
        return recs


def tok(value):
    """Flatten any nested value into a set of lowercased non-empty string tokens."""
    out = set()
    if value is None:
        return out
    if isinstance(value, str):
        s = value.strip().lower()
        if s:
            out.add(s)
    elif isinstance(value, list):
        for item in value:
            out |= tok(item)
    elif isinstance(value, dict):
        for item in value.values():
            out |= tok(item)
    else:
        out.add(str(value).lower())
    return out


def notes_str(rec):
    v = rec.get("applicationNotes")
    if isinstance(v, str):
        return v
    return "" if v is None else json.dumps(v)


def contact(rec):
    persons = rec.get("persons") or []
    if not persons:
        return None, None, rec.get("linkedin")
    p = persons[0]
    name = p.get("name") or {}
    full = " ".join(x for x in [name.get("firstName"), name.get("lastName")] if x) or None
    email = (p.get("emails") or {}).get("primaryEmail")
    linkedin = p.get("linkedinLink") or rec.get("linkedin")
    return full, email, linkedin


def baseline(validated):
    geo, lang, scope, skill = set(), set(), set(), set()
    for r in validated:
        geo |= tok(r.get("region")) | tok(r.get("country"))
        lang |= tok(r.get("languagesSpoken"))
        scope |= tok(r.get("partnerScope"))
        skill |= tok(r.get("skills"))
    return {"geo": geo, "lang": lang, "scope": scope, "skill": skill}


def score_one(rec, base):
    new_geo = sorted((tok(rec.get("region")) | tok(rec.get("country"))) - base["geo"])
    new_lang = sorted(tok(rec.get("languagesSpoken")) - base["lang"])
    new_scope = sorted(tok(rec.get("partnerScope")) - base["scope"])
    new_skill = sorted(tok(rec.get("skills")) - base["skill"])
    notes = notes_str(rec)
    proof = {
        "workspace_url": bool(PROOF_WORKSPACE.search(notes)),
        "customers": bool(PROOF_CUSTOMERS.search(notes)),
        "migration": bool(PROOF_MIGRATION.search(notes)),
    }
    has_proof = any(proof.values())
    score = (WEIGHTS["geo"] * len(new_geo)
             + WEIGHTS["lang"] * len(new_lang)
             + WEIGHTS["scope"] * len(new_scope)
             + WEIGHTS["skill"] * min(len(new_skill), WEIGHTS["skill_cap"])
             + (WEIGHTS["proof"] if has_proof else 0))
    if has_proof or score >= THRESHOLDS["A"]:
        tier = "A"
    elif score >= THRESHOLDS["B"]:
        tier = "B"
    else:
        tier = "C"
    name, email, linkedin = contact(rec)
    return {
        "name": rec.get("name"),
        "score": score,
        "tier": tier,
        "new_geo": new_geo,
        "new_lang": new_lang,
        "new_scope": new_scope,
        "new_skills": new_skill,
        "proof": {k: v for k, v in proof.items() if v},
        "team": rec.get("typeOfTeam"),
        "contact_name": name,
        "email": email,
        "linkedin": linkedin if isinstance(linkedin, str) else (linkedin or {}).get("primaryLinkUrl") if isinstance(linkedin, dict) else None,
        "website": (rec.get("website") or {}).get("primaryLinkUrl") if isinstance(rec.get("website"), dict) else None,
        "notes": notes.strip()[:400],
    }


def rank(recs):
    validated = [r for r in recs if r.get("validationStage") == "VALIDATED"]
    apps = [r for r in recs if r.get("validationStage") == "APPLICATION"]
    # Forward-compat: once callBookedAt exists, narrow to the un-booked (the chase set).
    has_booked_field = any("callBookedAt" in r for r in recs)
    if has_booked_field:
        apps = [r for r in apps if not r.get("callBookedAt")]
    base = baseline(validated)
    ranked = sorted((score_one(r, base) for r in apps), key=lambda d: -d["score"])
    return {
        "validated_count": len(validated),
        "application_count": len(apps),
        "booking_state_wired": has_booked_field,
        "coverage": {k: sorted(v) for k, v in base.items()},
        "ranked": ranked,
    }


def selftest():
    base_recs = [{"validationStage": "VALIDATED", "country": ["france"],
                  "languagesSpoken": ["french", "english"], "partnerScope": ["development"],
                  "skills": ["react", "postgres"]}]
    apps = [
        {"validationStage": "APPLICATION", "name": "gap+proof", "country": ["germany"],
         "languagesSpoken": ["german"], "applicationNotes":
         "Live workspace https://crm.acme.de — customers onboarded: Foo GmbH"},
        {"validationStage": "APPLICATION", "name": "skill-sprayer",
         "skills": ["php", "vue", "kotlin", "swift", "laravel", "mongodb"]},
        {"validationStage": "APPLICATION", "name": "empty"},
    ]
    out = rank(base_recs + apps)["ranked"]
    order = [r["name"] for r in out]
    assert order == ["gap+proof", "skill-sprayer", "empty"], order
    assert out[0]["tier"] == "A", out[0]
    assert out[1]["score"] == WEIGHTS["skill_cap"], out[1]  # 6 new skills capped at 3
    assert out[2]["tier"] == "C", out[2]
    print("selftest ok:", order)


if __name__ == "__main__":
    if "--selftest" in sys.argv:
        selftest()
    else:
        url, key = load_creds()
        print(json.dumps(rank(fetch_all_partners(url, key)), indent=2, ensure_ascii=False))
